### PR TITLE
509 Nihms Loader

### PR DESCRIPTION
**Pull request by tsande16**
This is the port of the fedora based data loader for nihms. This is the initial port that includes:

- Port to Pass-Support, keeping the behavior as-is
- Change of package name to org.eclipse.pass
- Hooking up to parent POM, Cleaning up the POM, updating some versions
- Switch from pass-model & pass-api to pass-core-object-service (passclient.newinstance())
- Separated integration test modules, but maybe consolidate them a bit
- Remove IDE warnings
- Potentially update credentials - needed to run this loader
- Verify tests, possibly rewrite
- Run ITs

To Manual Test:

`mvn clean install`

Will need API key: https://livejohnshopkins-my.sharepoint.com/:w:/g/personal/jgara1_jh_edu/EZog8wpGqz1Mt6DqTlX970sB74a6elEgguQ_uO0D-YtvfQ?e=4%3Abw4gqY&at=9

*Use most recent token that was generated

- Run The Data Harvester 
- Setup a config file

configfile.properties: 
```
# Format ought to be CSV, otherwise the loader won't be able to process the saved files
nihmsetl.api.url.param.format=csv

# Institution name, unclear as to how it is used
nihmsetl.api.url.param.inst=JOHNS HOPKINS UNIVERSITY

#  IPF (Institutional Profile File) number, the unique ID assigned to a grantee organization in the eRA system. 
nihmsetl.api.url.param.ipf=4134401

# The API token retrieved from the PACM website.  These expire every three months.
nihmsetl.api.url.param.api-token=PUT_API_TOKEN_HERE
```
Run command:

`java -Dnihmsetl.harvester.configfile=path_to_configfile.properties -jar nihms-data-harvest-cli-0.7.0-SNAPSHOT-shaded.jar1`

- This will output a /data directory in the same directory as the jar. If run successfully you will see three files:
    - compliant_nihmspubs_YYYYMMDDHHMMSS.csv
    - noncompliant_nihmspubs_YYYYMMDDHHMMSS.csv
    - inprocess_nihmspubs_YYYYMMDDHHMMSS.csv

- Once the data harvester is completed you can now run the transform load module
- Run pass-core jar
- Add grants to pass-core that are from the data file
- Set up transform load module configuration file:
```
nihmsetl.data.dir=path_to_harvested_files
nihmsetl.pmcurl.template=https://www.ncbi.nlm.nih.gov/pmc/articles/%s/
pass.core.url=http://localhost:8080
pass.core.user=backend
pass.core.password=moo
```
- Run command: `java -Dnihmsetl.loader.configfile=path_to_transform_load_config -jar nihms-data-transform-load-0.7.0-SNAPSHOT-shaded.jar`

- If successfully run you will see messages that the records have been successfully imported.

Examples of Grants to add:
```
{
  "data": {
      "type": "grant",
      "attributes": {
        "endDate": "2018-08-31T00:00:00.000Z",
        "awardNumber": "K23 DK119949",
        "awardDate": "2015-09-23T00:00:00.000Z",
        "awardStatus": "active",
        "projectName": "Reversible activation of critical period plasticity in visual cortex",
        "startDate": "2015-09-01T00:00:00.000Z",
        "localKey": "johnshopkins.edu:grant:12345"
      }
  }
}


{
  "data": {
      "type": "grant",
      "attributes": {
        "endDate": "2018-08-31T00:00:00.000Z",
        "awardNumber": "R01 DK108304",
        "awardDate": "2015-09-23T00:00:00.000Z",
        "awardStatus": "active",
        "projectName": "Reversible activation of critical period plasticity in visual cortex",
        "startDate": "2015-09-01T00:00:00.000Z",
        "localKey": "johnshopkins.edu:grant:12345"
      }
  }
}



{
  "data": {
      "type": "grant",
      "attributes": {
        "endDate": "2018-08-31T00:00:00.000Z",
        "awardNumber": "P30 CA006973",
        "awardDate": "2015-09-23T00:00:00.000Z",
        "awardStatus": "active",
        "projectName": "Reversible activation of critical period plasticity in visual cortex",
        "startDate": "2015-09-01T00:00:00.000Z",
        "localKey": "johnshopkins.edu:grant:12345"
      }
  }
}


{
  "data": {
      "type": "grant",
      "attributes": {
        "endDate": "2018-08-31T00:00:00.000Z",
        "awardNumber": "U54 CA153710",
        "awardDate": "2015-09-23T00:00:00.000Z",
        "awardStatus": "active",
        "projectName": "Reversible activation of critical period plasticity in visual cortex",
        "startDate": "2015-09-01T00:00:00.000Z",
        "localKey": "johnshopkins.edu:grant:12345"
      }
  }
}


{
  "data": {
      "type": "grant",
      "attributes": {
        "endDate": "2018-08-31T00:00:00.000Z",
        "awardNumber": "R61 HD104570",
        "awardDate": "2015-09-23T00:00:00.000Z",
        "awardStatus": "active",
        "projectName": "Reversible activation of critical period plasticity in visual cortex",
        "startDate": "2015-09-01T00:00:00.000Z",
        "localKey": "johnshopkins.edu:grant:12345"
      }
  }
}

{
  "data": {
      "type": "grant",
      "attributes": {
        "endDate": "2018-08-31T00:00:00.000Z",
        "awardNumber": "R01 DE014581",
        "awardDate": "2015-09-23T00:00:00.000Z",
        "awardStatus": "active",
        "projectName": "Reversible activation of critical period plasticity in visual cortex",
        "startDate": "2015-09-01T00:00:00.000Z",
        "localKey": "johnshopkins.edu:grant:12345"
      }
  }
}

{
  "data": {
      "type": "grant",
      "attributes": {
        "endDate": "2018-08-31T00:00:00.000Z",
        "awardNumber": "U01 DE018993",
        "awardDate": "2015-09-23T00:00:00.000Z",
        "awardStatus": "active",
        "projectName": "Reversible activation of critical period plasticity in visual cortex",
        "startDate": "2015-09-01T00:00:00.000Z",
        "localKey": "johnshopkins.edu:grant:12345"
      }
  }
}

{
  "data": {
      "type": "grant",
      "attributes": {
        "endDate": "2018-08-31T00:00:00.000Z",
        "awardNumber": "P41 EB024495",
        "awardDate": "2015-09-23T00:00:00.000Z",
        "awardStatus": "active",
        "projectName": "Reversible activation of critical period plasticity in visual cortex",
        "startDate": "2015-09-01T00:00:00.000Z",
        "localKey": "johnshopkins.edu:grant:12345"
      }
  }
}

{
  "data": {
      "type": "grant",
      "attributes": {
        "endDate": "2018-08-31T00:00:00.000Z",
        "awardNumber": "F31 NR019205",
        "awardDate": "2015-09-23T00:00:00.000Z",
        "awardStatus": "active",
        "projectName": "Reversible activation of critical period plasticity in visual cortex",
        "startDate": "2015-09-01T00:00:00.000Z",
        "localKey": "johnshopkins.edu:grant:12345"
      }
  }
}

{
  "data": {
      "type": "grant",
      "attributes": {
        "endDate": "2018-08-31T00:00:00.000Z",
        "awardNumber": "K23 MH118431",
        "awardDate": "2015-09-23T00:00:00.000Z",
        "awardStatus": "active",
        "projectName": "Reversible activation of critical period plasticity in visual cortex",
        "startDate": "2015-09-01T00:00:00.000Z",
        "localKey": "johnshopkins.edu:grant:12345"
      }
  }
}

{
  "data": {
      "type": "grant",
      "attributes": {
        "endDate": "2018-08-31T00:00:00.000Z",
        "awardNumber": "K23 MH118431",
        "awardDate": "2015-09-23T00:00:00.000Z",
        "awardStatus": "active",
        "projectName": "Reversible activation of critical period plasticity in visual cortex",
        "startDate": "2015-09-01T00:00:00.000Z",
        "localKey": "johnshopkins.edu:grant:12345"
      }
  }
}
```

### CODE REVIEW

**Comment by markpatton** [comment]
Weirdly, the tests aren't being run on the pr.

A few checkstyle errors to fix:
```
[ERROR] /home/msp/work/pass/pass-support/pass-nihms-loader/nihms-etl-util/src/main/java/org/eclipse/pass/loader/nihms/util/ConfigUtil.java:22:8: Unused import - java.net.URI. [UnusedImports]
[ERROR] /home/msp/work/pass/pass-support/pass-nihms-loader/nihms-etl-util/src/main/java/org/eclipse/pass/loader/nihms/util/ConfigUtil.java:23:8: Unused import - java.net.URISyntaxException. [UnusedImports]
[ERROR] /home/msp/work/pass/pass-support/pass-nihms-loader/nihms-etl-util/src/test/java/org/eclipse/pass/loader/nihms/util/FileUtilTest.java:31:1: Import statement for 'org.junit.jupiter.api.Assertions.*' is in the wrong order. Should be in the 'STATIC' group, expecting not assigned imports on this line. [CustomImportOrder]
[ERROR] /home/msp/work/pass/pass-support/pass-nihms-loader/nihms-etl-util/src/test/java/org/eclipse/pass/loader/nihms/util/FileUtilTest.java:31:47: Using the '.*' form of import should be avoided - org.junit.jupiter.api.Assertions.*. [AvoidStarImport]
```

File: pass-data-client/src/main/java/org/eclipse/pass/support/client/model/AggregatedDepositStatus.java [hunk 1]
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright 2018 Johns Hopkins University
+ * Copyright 2023 Johns Hopkins University
```
Comment (rpoet-jh) [inline pass-data-client/src/main/java/org/eclipse/pass/support/client/model/AggregatedDepositStatus.java:2]: Given the recent decision to change the copyright statement structure in our project, and that the year update is the only change in the pass-data-client module files, I'd revert this change just to limit the number of files changed in this PR.  Also, I think I remember hearing that the year doesn't have to be updated even with the new copyright statement structure.

File: pass-nihms-loader/nihms-pass-client/src/main/java/org/eclipse/pass/client/nihms/NihmsPassClientService.java [hunk 2]
```diff
@@ -0,0 +1,659 @@
...
+import org.eclipse.pass.support.client.model.Submission;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * NIHMS PASS client service deals with interactions with the data via the PASS client and controls local data caches
+ *
+ * @author Karen Hanson
+ */
+public class NihmsPassClientService {
```
Comment (rpoet-jh) [inline pass-nihms-loader/nihms-pass-client/src/main/java/org/eclipse/pass/client/nihms/NihmsPassClientService.java:53]: This class is pretty long.  Consider refactoring into maybe a read and write service.

File: pass-nihms-loader/nihms-pass-client/src/main/java/org/eclipse/pass/client/nihms/cache/GrantIdCache.java [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+package org.eclipse.pass.client.nihms.cache;
+
+import java.util.HashMap;
+
+/**
+ * Caches awardNumber and grantId combination for easy lookup
+ *
+ * @author Karen Hanson
+ */
+public class GrantIdCache {
```
Comment (rpoet-jh) [inline pass-nihms-loader/nihms-pass-client/src/main/java/org/eclipse/pass/client/nihms/cache/GrantIdCache.java:25]: I don't see the purpose of these xxxCache classes.  They are thin wrappers around a Map for caching purposes.  I would ask why not just use a Map in `NihmsPassClientService`, etc.  If the Map really needs to be synchronized, `Collections.synchronizedMap()` or `ConcurrentHashMap` could be used.

**Review by rpoet-jh — COMMENTED** [review]
Great job, Tim!  I have some comments.  I realize that some of my comments are on code that was just ported over, but I thought I would make the comment, and we can decide whether to make any changes or not.

File: pass-data-client/src/main/java/org/eclipse/pass/support/client/model/AggregatedDepositStatus.java (hunk 1 above)
Comment (tsande16) [reply pass-data-client/src/main/java/org/eclipse/pass/support/client/model/AggregatedDepositStatus.java:2]: Done.

File: pass-nihms-loader/entrez-pmid-lookup/pom.xml [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+    <dependency>
+      <groupId>org.junit.jupiter</groupId>
+      <artifactId>junit-jupiter-engine</artifactId>
+      <version>${junit.jupiter.version}</version>
+    </dependency>
+
+    <dependency>
+      <groupId>ch.qos.logback</groupId>
+      <artifactId>logback-classic</artifactId>
+      <scope>test</scope>
```
[the remark this answers is not in the document]
Comment (tsande16) [reply pass-nihms-loader/entrez-pmid-lookup/pom.xml:68]: Done.

File: pass-nihms-loader/nihms-data-harvest-cli/dependency-reduced-pom.xml [hunk 5]
```diff
@@ -0,0 +1,48 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
[the remark this answers is not in the document]
Comment (tsande16) [reply pass-nihms-loader/nihms-data-harvest-cli/dependency-reduced-pom.xml:1]: Done.

File: pass-nihms-loader/nihms-data-harvest/src/main/java/org/eclipse/pass/loader/nihms/FileWatcher.java [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+import java.util.Arrays;
+import java.util.Optional;
+
+/**
+ * Utility to poll a directory to check a file is downloaded. Once the file appears it
+ * sends the new filepath back
+ *
+ * @author Karen Hanson
+ */
+public class FileWatcher {
```
[the remark this answers is not in the document]
Comment (tsande16) [reply pass-nihms-loader/nihms-data-harvest/src/main/java/org/eclipse/pass/loader/nihms/FileWatcher.java:29]: Not used. Checked git history, unsure if this was ever used, perhaps an anticipated feature. Checked with Russ and decided to remove. 

Done.

File: pass-nihms-loader/nihms-data-harvest/src/main/java/org/eclipse/pass/loader/nihms/NihmsHarvesterConfig.java [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+import java.util.Map;
+
+import org.eclipse.pass.loader.nihms.util.ConfigUtil;
+
+/**
+ * Holds information and methods required to configure the NIHMS harvester tool.
+ *
+ * @author Karen Hanson
+ */
+public class NihmsHarvesterConfig {
```
[the remark this answers is not in the document]
Comment (tsande16) [reply pass-nihms-loader/nihms-data-harvest/src/main/java/org/eclipse/pass/loader/nihms/NihmsHarvesterConfig.java:28]: Discussed with Russ and decided to move to cleanup ticket: https://github.com/eclipse-pass/main/issues/602

Will use application.yml + `@ConfigurationProperties`

File: pass-nihms-loader/nihms-pass-client/src/main/java/org/eclipse/pass/client/nihms/NihmsPassClientService.java [hunk 8]
```diff
@@ -0,0 +1,659 @@
...
+     *
+     * @param issn the issn
+     * @return the journal ID for the ISSN, may be {@code null} if not found
+     * @throws IOException if there is an error reading the journal
+     */
+    public String findJournalByIssn(String issn) throws IOException {
+        if (nullOrEmpty(issn)) {
+            return null;
+        }
+        //TODO: equals needs to change to hasmember
```
[the remark this answers is not in the document]
Comment (tsande16) [reply pass-nihms-loader/nihms-pass-client/src/main/java/org/eclipse/pass/client/nihms/NihmsPassClientService.java:414]: Done.

File: pass-nihms-loader/nihms-etl-util/src/main/java/org/eclipse/pass/loader/nihms/util/ProcessingUtil.java [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+
+    private static final String DEFAULT_DATE_PATTERN = "yyyy-MM-dd";
+
+    /**
+     * Returns true if a string is empty or null
+     *
+     * @param str the string, may be {@code null}
+     * @return true if a string is empty or null
+     */
+    public static boolean nullOrEmpty(String str) {
```
[the remark this answers is not in the document]
Comment (tsande16) [reply pass-nihms-loader/nihms-etl-util/src/main/java/org/eclipse/pass/loader/nihms/util/ProcessingUtil.java:43]: Done.

[the remark this answers is not in the document]
Comment (tsande16) [reply pass-nihms-loader/nihms-etl-util/src/main/java/org/eclipse/pass/loader/nihms/util/ProcessingUtil.java:43]: @rpoet-jh Actually not sure if this done. Did you want me to replace only the method nullOrEmpty() or all the methods in the ProcessingUtil class? The formatDate() is doing a little more, so it might be harder to replace.

File: pass-nihms-loader/nihms-pass-client/src/main/java/org/eclipse/pass/client/nihms/NihmsPassClientService.java (hunk 2 above)
Comment (tsande16) [reply pass-nihms-loader/nihms-pass-client/src/main/java/org/eclipse/pass/client/nihms/NihmsPassClientService.java:53]: Yes, I agree, its a bit lengthy. I can address this now or add it to the clean up?

File: pass-nihms-loader/nihms-pass-client/src/main/java/org/eclipse/pass/client/nihms/cache/GrantIdCache.java (hunk 3 above)
Comment (tsande16) [reply pass-nihms-loader/nihms-pass-client/src/main/java/org/eclipse/pass/client/nihms/cache/GrantIdCache.java:25]: I believe we created a ticket to address this a while back: https://github.com/eclipse-pass/main/issues/566

Do you want me to address it in that ticket or add it to the pass-support nihms cleanup ticket? 

https://github.com/eclipse-pass/main/issues/602

File: pass-nihms-loader/nihms-etl-util/src/main/java/org/eclipse/pass/loader/nihms/util/ProcessingUtil.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (rpoet-jh) [reply pass-nihms-loader/nihms-etl-util/src/main/java/org/eclipse/pass/loader/nihms/util/ProcessingUtil.java]: Just the string operation methods, so it's done.

File: pass-nihms-loader/nihms-pass-client/src/main/java/org/eclipse/pass/client/nihms/NihmsPassClientService.java (hunk 2 above)
Comment (rpoet-jh) [reply pass-nihms-loader/nihms-pass-client/src/main/java/org/eclipse/pass/client/nihms/NihmsPassClientService.java:53]: It's up to you, if it would be quick (couple hours), may be worth doing now.

File: pass-nihms-loader/nihms-pass-client/src/main/java/org/eclipse/pass/client/nihms/cache/GrantIdCache.java (hunk 3 above)
Comment (rpoet-jh) [reply pass-nihms-loader/nihms-pass-client/src/main/java/org/eclipse/pass/client/nihms/cache/GrantIdCache.java:25]: I'd say to probably combine 566 into 602, just so all items are in one ticket.

**Review by markpatton — APPROVED** [review]
Looks good. I was able to follow the instructions and successfully harvest and load data. I had to increase the timeout to 30 seconds in order to do harvesting. Maybe that would be a good default? During the load there were a bunch of log messages about a missing publication id. I'm not sure if that is significant.

For the code, there are a few cleanups that would be good to do, but they could be done later. There are a bunch of classes to do caching that are not really needed. ConcurrentHashMap could just be used instead. Also many of the modules are just a handful of classes. You might consider refactoring into fewer modules. For example one core module and two cli modules.

The nihms-data-harvest-cli module produces the harvest command line app. The nihms-data-transform-load module produces the load command line app. We should rename nihms-data-transform-load to something like nihms-data-load-cli to be consistent.

File: pass-nihms-loader/nihms-data-transform-load/src/main/java/org/eclipse/pass/loader/nihms/NihmsTransformLoadApp.java [hunk 10]
```diff
@@ -0,0 +1,112 @@
...
+    private static Logger LOG = LoggerFactory.getLogger(NihmsTransformLoadApp.class);
+
+    private static final String NIHMS_CONFIG_FILEPATH_PROPKEY = "nihmsetl.loader.configfile";
+    private static final String DEFAULT_CONFIG_FILENAME = "nihms-loader.properties";
+
+    /**
+     * These are the only system properties that can be loaded in from the properties file.
+     * Existing values will not be overwritten, these will just be added if missing.
+     */
+    private static final String[] SYSTEM_PROPERTIES = {"pass.fedora.user", "pass.fedora.password",
```
[the remark this answers is not in the document]
Comment (tsande16) [reply pass-nihms-loader/nihms-data-transform-load/src/main/java/org/eclipse/pass/loader/nihms/NihmsTransformLoadApp.java]: Done.

File: pass-nihms-loader/nihms-etl-util/src/main/java/org/eclipse/pass/loader/nihms/util/ProcessingUtil.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (tsande16) [reply pass-nihms-loader/nihms-etl-util/src/main/java/org/eclipse/pass/loader/nihms/util/ProcessingUtil.java]: Discussed with Russ, instead of using isBlank(), will use isEmpty() and use isNotEmpty instead of !isEmpty().

File: pass-nihms-loader/nihms-data-transform-load/src/main/java/org/eclipse/pass/loader/nihms/NihmsTransformLoadService.java [hunk 11]
```diff
@@ -0,0 +1,227 @@
...
+    private PmidLookup pmidLookup;
+
+    private static CompletedPublicationsCache completedPubsCache;
+
+    private SubmissionStatusService statusService;
+
+    /**
+     * Default constructor for NihmsTransformLoadService
+     */
+    public NihmsTransformLoadService() {
```
[the remark this answers is not in the document]
Comment (tsande16) [reply pass-nihms-loader/nihms-data-transform-load/src/main/java/org/eclipse/pass/loader/nihms/NihmsTransformLoadService.java:60]: Will move to cleanup ticket.

File: pass-nihms-loader/nihms-data-transform-load/src/main/java/org/eclipse/pass/loader/nihms/CompletedPublicationsCache.java [hunk 12]
```diff
@@ -0,0 +1,144 @@
...
+                                                         FileUtil.getCurrentDirectory() + CACHEPATH_DEFAULT);
+        cacheFile = new File(sCacheFile);
+        loadFromFile();
+    }
+
+    /**
+     * Get singleton instance of cache
+     * @return the singleton instance
+     */
+    public static synchronized CompletedPublicationsCache getInstance() {
```
[the remark this answers is not in the document]
Comment (tsande16) [reply pass-nihms-loader/nihms-data-transform-load/src/main/java/org/eclipse/pass/loader/nihms/CompletedPublicationsCache.java:61]: Will move to cleanup ticket.

File: pass-nihms-loader/nihms-data-transform-load/src/main/java/org/eclipse/pass/loader/nihms/SubmissionDTO.java [hunk 13]
```diff
@@ -0,0 +1,149 @@
...
+import org.eclipse.pass.support.client.model.RepositoryCopy;
+import org.eclipse.pass.support.client.model.Submission;
+
+/**
+ * Data transfer object to hold the various components of a NIHMS Submission up to the point of
+ * update or create.
+ *
+ * @author Karen Hanson
+ */
+public class SubmissionDTO {
```
[the remark this answers is not in the document]
Comment (tsande16) [reply pass-nihms-loader/nihms-data-transform-load/src/main/java/org/eclipse/pass/loader/nihms/SubmissionDTO.java:28]: Will move to cleanup ticket.

File: pass-nihms-loader/nihms-data-transform-load/src/test/java/org/eclipse/pass/loader/nihms/TransformAndLoadInProcess.java [hunk 14]
```diff
@@ -0,0 +1,286 @@
...
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.ExtendWith;
+import org.mockito.junit.jupiter.MockitoExtension;
+
+
+/**
+ * @author Karen Hanson
+ */
+@ExtendWith(MockitoExtension.class)
+public class TransformAndLoadInProcess extends NihmsSubmissionEtlITBase {
```
[the remark this answers is not in the document]
Comment (tsande16) [reply pass-nihms-loader/nihms-data-transform-load/src/test/java/org/eclipse/pass/loader/nihms/TransformAndLoadInProcess.java:52]: Done.
